UI Improvement and Features

WindingMotor opened the pull request:
# UI Improvement and Features

This PR adds significant UI changes and several new features to improve the user experience and functionality.

## New Features:
- Path Length Feedback: Added real-time feedback in RuntimeDisplay to indicate path length changes.
- 0.5m x 0.5m Grid Option: New grid option available in Editor Settings.
- Robot Details Display: Added section in Editor Settings showing current robot angle and position.
- Enhanced Tree Card Visuals: Added icons to tree cards and robot path items for better visual distinction.
- Mini Info Cards: Introduced quick-glance information for rotation and event marker tree cards.
- Manual Entry for Markers: Enabled manual position entry for event markers and rotation targets.
- Wait Time Command Controls: Added increment/decrement buttons for wait time commands.

## UI Improvements:
- Redesigned Sidebar: Reworked layout for a more compact and efficient design.
- Top Bar Optimization: Decreased overall size for better screen real estate usage.
- Event Marker Visuals: Enhanced appearance on the path.
- Icon Updates: Improved icon choices and colors for clarity and consistency.
- Font Size Adjustments: Reduced font size for checkbox labels to improve readability.
- Button Refinements: Streamlined various buttons and relocated others for better usability.
- Replay Bar Enhancements: Improved layout and added a replay button.
- Holonomic Mode Indicator: Added visual indicator for holonomic mode status.
- Custom Color Scheme: Enhanced theming for greater visual consistency.
- Lock Animation: Added animation effects for path editor locking.
- Constraint Zone Visibility: Increased stroke thickness for better visibility.
- Additional Tooltips: Expanded tooltip coverage for improved user guidance.
- Arrow Key Functionality: Added 45-degree increment/decrement for rotation target entry using arrow keys.
- Graph Enhancements: Improved visuals and added hover information.
- Loading Experience: Added helpful checklist and IP status during loading.
- Sidebar icon updates much better than before

---

As this is my first public pull request, I would greatly appreciate feedback. I've also included several before/after images demonstrating the UI changes.

## Before & After Images


### Path Editor 
![PathEditorBEFORE](https://github.com/user-attachments/assets/f327468e-5872-45f4-852a-b3824500a104)

![PathEditorAFTER](https://github.com/user-attachments/assets/35714c71-88a8-4541-9f4a-e27c585f73b1)


### Sidebar

![ProjectBrowserBEFORE](https://github.com/user-attachments/assets/550909d9-4a25-4027-864d-ee6740b3a74b)

![ProjectBrowserAFTER](https://github.com/user-attachments/assets/53f4af26-9faa-4175-9ffd-e0cb32a94f3d)

### Telemetry 

![TelemetryBEFORE](https://github.com/user-attachments/assets/927a28a7-f108-4435-b5ae-dd1e39b80671)

![TelemetryAFTER](https://github.com/user-attachments/assets/7c8ce116-6d24-45e9-ae2c-deebb579b2dd)

### Grid & Robot Details

![PathEditorGridAndRobotDetails](https://github.com/user-attachments/assets/50b80052-11db-41e5-acd5-cb17d85a5d1d)

### Waypoints Tree

![PathEditorWaypointsBEFORE](https://github.com/user-attachments/assets/36fa7889-2188-4d1f-87cb-844f52ed2ba9)

![PathEditorWaypointsAFTER](https://github.com/user-attachments/assets/f5a73283-22e5-41b3-88ab-631e181bd958)

### Rotation Target Tree

![PathEditorRotationTargetsBEFORE](https://github.com/user-attachments/assets/b3bac191-ed68-4222-9950-b7cd19431086)

![PathEditorRotationTargetsAFTER](https://github.com/user-attachments/assets/64f31f28-d4ec-4964-99d6-6c30889c7c61)

---

mjansen4857 commented:
A lot of this looks really awesome, thanks! Just gonna have to let this sit here for a little bit though as I have another PR that has a bunch of stuff I need to get merged first before I can work on or merge anything else. I don't think it will break too much here though.

Once I get that stuff finished, tested, and merged I can come back to review this. Sorry for the delay.

mjansen4857 commented:
Ok that PR is merged so you can take a look at those merge conflicts. I haven't reviewed in depth but from an overview perspective there's a few things I have feedback on.

* The side drawer should remain an actual drawer with the same styling it had previously. Its current usage follows the material design guidelines which helps keep the whole app feeling cohesive. I think a way to improve the current design would be to make the telemetry icon change that you did, by making the icon for the telemetry page the connection status indicator. Then, you could move the change project button down in to the corner that the indicator used to be in to clean things up a bit.
* I don't think the car icon and the arrow instead of the dot for the robot preview are necessary. Once you merge you'll see that swerve modules get represented now so the car icon isn't needed. Also, keeping the dot instead of the arrow just keeps all the representations of the robot looking similar.
* I think the graph changes on the telemetry page are mostly fine, but I think the changes to include axis labels and the different grid make it look more cluttered. I think leaving the grid/axis as-is would look a lot cleaner and having the ability to mouse over to see a value would satisfy anyone who wants to know actual values. At the end of the day, AdvantageScope is a better tool for this anyways, this is just a quick and dirty way to see that telemetry for people who don't use it.
* Somewhat minor thing but I think the extra info display in the goal end state should also show the velocity.
* I'm not sure if there's really a purpose in having that holonomic mode indicator. Users should know which mode they are in based on the look and function of the editor.
* I think the add folder/path buttons in the project browser should remain filled icon buttons with the different colors. It just gives it some more contrast and draws people to them. Making the add path button less visible could cause confusion for some people since its not as obvious how they can create a new path.

Overall, the rest looks pretty great.

WindingMotor commented:
Ill take a look at the merge conflicts and make the changes you suggested. Thanks for the feedback!

WindingMotor commented:
I have merged and made the improvements you mentioned earlier. I have also fixed the app icon not working on Linux systems and added a path/auto search bars on the Project Browser. I'm also loving the new swerve modules visualization!
Here are some images for reference: 

**Project Browser:** 
![image](https://github.com/user-attachments/assets/d0932d96-c5d6-4cab-a9c3-96f6e1863166)

**Side Drawer:**
![image](https://github.com/user-attachments/assets/0d8051ac-3c27-4216-b062-973af2dcf71a)

**Telemetry Page:**
![image](https://github.com/user-attachments/assets/fe6cf227-9003-4d26-a965-cb70d8ee8129)

**Path Editor:**
![image](https://github.com/user-attachments/assets/1a051ae3-a9dc-432f-b778-8349d35ab361)

mjansen4857 commented:
Looking good. Can you add tests for the new stuff you added? Most things you touched should already be covered, but testing the search bars, the extra info displays in the path tree, and the position text field for event markers/rotation targets would be great. There's lots of examples in the tests already there so it should hopefully be easy to figure out but if you need any help let me know.

It also looks like a few tests are failing so you'll need to look into why they're not working.

WindingMotor commented:
Sounds good, I will get to work on those tests.

WindingMotor commented:
After taking a quick look it appears .mocks.dart files are included in .gitignore. So my flutter files are throwing errors.  I am new to using tests in flutter so any help would be appreciated, thanks! 

![image](https://github.com/user-attachments/assets/06605f62-e75c-45a7-ac88-645c4056d43d)

mjansen4857 commented:
The mocks need to be generated using `dart run build_runner build`. You only have to do this once unless they get changed.

WindingMotor commented:
It took a few days but i have merged with your recent commits and updated all the old tests while adding the new ones. I have also made a few UI changes.

**Project page:**
![image](https://github.com/user-attachments/assets/6d521939-7598-4f51-a8a8-832f824783ed)

**Failed to generate snackbar:**
![image](https://github.com/user-attachments/assets/5696f0ab-0578-414d-bf10-76cdc852ad58)

mjansen4857 commented:
It looks like a lot of coverage is missing for some of the tests. From looking at the coverage report it looks like most of it was previously covered before this PR but now isn't. Perhaps you deleted some tests and never added them back? telemetry_page.dart specifically looks like it lost a lot of coverage. Try seeing if you can get those tests back and we'll see what the project coverage looks like after that. I usually try to aim for not letting the total project coverage decrease.

mjansen4857 commented on lib/widgets/editor/tree_widgets/commands/wait_command_widget.dart:
```diff
@@ -20,37 +20,63 @@ class WaitCommandWidget extends StatelessWidget {
     this.onDuplicateCommand,
   });
 
+  void _increaseWaitTime() {
+    _updateWaitTime(command.waitTime + 0.2);
+  }
+
+  void _decreaseWaitTime() {
+    _updateWaitTime(command.waitTime - 0.2);
```
The `NumberTextField` widget that the wait command uses already has an increment feature using the arrow keys. Its a bit less visible than this but it seems a little redundant to have it increment in 2 different places. Also, an increment of 0.1s is probably a lot better for fine control and quick adjustments.

WindingMotor replied:
Yeah i guess its quite a bit redundant

mjansen4857 commented on lib/widgets/editor/tree_widgets/editor_settings_tree.dart:
```diff
@@ -123,4 +109,54 @@ class _EditorSettingsTreeState extends State<EditorSettingsTree> {
+    );
+  }
+
+  void _updateSetting(String key, bool? value) {
+    if (value != null) {
+      setState(() {
+        switch (key) {
+          case PrefsKeys.snapToGuidelines:
+            _snapToGuidelines = value;
+            break;
+          case PrefsKeys.hidePathsOnHover:
+            _hidePathsOnHover = value;
+            break;
+          case PrefsKeys.showStates:
+            _showStates = value;
+            break;
+          case PrefsKeys.showRobotDetails:
+            _showRobotDetails = value;
+            break;
+          case PrefsKeys.showGrid:
+            _showGrid = value;
+            break;
+        }
+        _prefs.setBool(key, value);
+      });
+    }
+  }
```
Not really a fan of using a switch statement like this here. I'd probably just replace every usage of this method with something like
```
setState(() {
    _snapToGuidelines = value!;
});
_prefs.setBool(PrefsKeys.snapToGuidelines, _snapToGuidelines);
```
You can safely use the `!` operator since these are not 3 state checkboxes and will never return null. Just makes things a bit cleaner and easier to read.

WindingMotor replied:
Yeah that would make it quite easier to read

mjansen4857 commented on lib/widgets/editor/tree_widgets/path_tree.dart:
```diff
@@ -34,6 +34,7 @@ class PathTree extends StatefulWidget {
   final num? pathRuntime;
   final bool holonomicMode;
   final PathConstraints defaultConstraints;
+  final num? previousPathRuntime;
 
```
Perhaps instead of taking in both `pathRuntime` and `previousPathRuntime` now it could just take in a nullable runtime display widget? Then all this widget has to do is display that runtime widget if its provided. Then it shouldn't need the `didUpdateWidget` stuff either. Same thing goes for the auto tree.

WindingMotor replied:
Good idea, made it so split_path_editor has the RuntimeDisplay object and that is used for the PathTree instead.

mjansen4857 commented on lib/widgets/editor/tree_widgets/tree_card_node.dart:
```diff
@@ -2,6 +2,7 @@ import 'package:flutter/material.dart';
 
 class TreeCardNode extends StatelessWidget {
   final Widget title;
+  final Widget? icon;
```
Should probably be renamed to `leading` to match other flutter widget conventions since this doesn't strictly have to be an icon.

WindingMotor replied:
Fair point, updated it to `leading` and all of its references

mjansen4857 commented on lib/widgets/editor/tree_widgets/waypoints_tree.dart:
```diff
@@ -151,15 +159,32 @@ class _WaypointsTreeState extends State<WaypointsTree> {
           Tooltip(
             message: waypoint.isLocked ? 'Unlock' : 'Lock',
             waitDuration: const Duration(seconds: 1),
-            child: IconButton(
-              onPressed: () {
-                waypoint.isLocked = !waypoint.isLocked;
+            child: GestureDetector(
+              onTap: () {
+                setState(() {
+                  waypoint.isLocked = !waypoint.isLocked;
+                });
                 widget.onPathChanged?.call();
               },
-              icon: Icon(waypoint.isLocked ? Icons.lock : Icons.lock_open,
-                  color: colorScheme.onSurface),
+              child: AnimatedSwitcher(
+                duration: const Duration(milliseconds: 300),
+                transitionBuilder: (Widget child, Animation<double> animation) {
+                  return ScaleTransition(scale: animation, child: child);
+                },
+                child: Icon(
+                  waypoint.isLocked
+                      ? Icons.lock_rounded
+                      : Icons.lock_open_rounded,
+                  key: ValueKey<bool>(waypoint.isLocked),
+                  color: waypoint.isLocked
+                      ? colorScheme.primary
+                      : colorScheme.onSurface,
+                  size: 20,
+                ),
+              ),
```
I'm pretty sure there's a way to use an animated icon with `IconButton`. `IconButton` takes in a `Widget` for the icon so you should be able to just pass in the `AnimatedSwitcher`. Using a gesture detector instead of an actual button sketches me out a little lol. It also has an ink well so it reacts to clicks and stuff.

WindingMotor replied:
A bit of an oversight, its been updated now to use a normal icon button with an AnimatedSwitcher as the icon

mjansen4857 commented on lib/widgets/my_application.cc:
```diff
@@ -0,0 +1,107 @@
+#include "my_application.h"
```
Pretty sure this file is here by mistake. There should only be dart code in the lib folder.

WindingMotor replied:
Oops, its been removed now

mjansen4857 reviewed (changes requested):
Still need to actually run the build and see if I have any feedback for how it looks/works, but from a code perspective it looks pretty good besides a few minor suggestions.

WindingMotor commented:
I will go ahead and take a look at your suggestions and codecov's report. Hopefully we can get that coverage back up.

WindingMotor commented on lib/widgets/editor/split_path_editor.dart:
```diff
@@ -748,10 +749,27 @@ class _SplitPathEditorState extends State<SplitPathEditor>
+            content: Text(
+              'Failed to generate trajectory. Try adjusting the path shape or the positions of rotation targets',
+              style: TextStyle(
+                  color: Theme.of(context).colorScheme.onSurfaceVariant),
+            ),
+            backgroundColor: Theme.of(context).colorScheme.surface,
```
Edited it to use onErrorContainer to make it stand out more on both

WindingMotor commented on lib/widgets/editor/split_auto_editor.dart:
```diff
@@ -309,10 +310,28 @@ class _SplitAutoEditorState extends State<SplitAutoEditor>
+          content: Text(
+            'Failed to generate trajectory for ${widget.auto.name}. Try adjusting the path shape or the positions of rotation targets.',
+            style: TextStyle(
+                color: Theme.of(context).colorScheme.onSurfaceVariant),
+          ),
+          backgroundColor: Theme.of(context).colorScheme.surface,
```
Same as below

WindingMotor commented on lib/widgets/editor/path_painter.dart:
```diff
@@ -98,8 +103,12 @@ class PathPainter extends CustomPainter {
+          paths[i],
+          canvas,
+          (hoveredPath == paths[i].name)
+              ? Theme.of(context).colorScheme.primary
```
Yeah, its been reverted back to orange

WindingMotor commented on lib/widgets/editor/path_painter.dart:
```diff
@@ -13,6 +13,7 @@ import 'package:pathplanner/util/path_painter_util.dart';
 import 'package:shared_preferences/shared_preferences.dart';
 
 class PathPainter extends CustomPainter {
+  final BuildContext context;
```
Good idea

WindingMotor commented on lib/pages/nav_grid_page.dart:
```diff
@@ -134,7 +134,7 @@ class _NavGridPageState extends State<NavGridPage> {
         ),
         Align(
-          alignment: Alignment.bottomRight,
+          alignment: Alignment.bottomLeft,
```
Yup

WindingMotor commented on lib/pages/project/project_page.dart:
```diff
@@ -287,7 +298,7 @@ class _ProjectPageState extends State<ProjectPage> {
         ),
         Align(
-          alignment: Alignment.bottomRight,
+          alignment: Alignment.bottomLeft,
```
Same as below

WindingMotor commented on lib/pages/project/project_page.dart:
```diff
@@ -413,26 +424,12 @@ class _ProjectPageState extends State<ProjectPage> {
           margin: const EdgeInsets.all(0),
-          color: colorScheme.primary,
-          surfaceTintColor: colorScheme.surfaceTint,
+          surfaceTintColor: colorScheme.surface,
```
Added

WindingMotor commented:
A test slipped one of my checks. Ill get to work on it and try to improve coverage.

WindingMotor commented:
Alright i made some improvements to the tests on the path tree, telemetry page, and waypoints tree.

mjansen4857 commented:
So I think a couple things are still hurting your coverage:
1. In the telemetry page tests you never give it data to display when connected. [See the lines that were removed here](https://github.com/mjansen4857/pathplanner/pull/705/files#diff-0b6292b03c4a9c24977f6a69df3649a04479614e4619f7c1579e3c602d6bd9c6R107-R108)
2. I'm not 100% sure but I think sharing one telemetry mock across all tests can cause issues because the same object will get different mock responses applied to it by all the tests. You should change that back to each telemetry test creating its own telemetry object.
3. The 'showDetails' flag for drawing the robot outline in path_painter_util is always false in tests. Set the initial setting value for showing details to true in split_path_editor_test. You'll see the block of settings stuff at the end of the `setUp` method.

Addressing those things should probably be enough to get the coverage back up to a good level.

WindingMotor commented:
Okay, i updated the telemetry page test and my local test coverage says its pretty good. I've also fixed the showDetails flag.

mjansen4857 commented:
Ok so the fix I gave you for painting path details didn't work because in tests it never gets a trajectory generated because that happens asynchronously. The real fix here is to add that default setting to true in split_choreo_path_editor_test since that will actually draw trajectories. I've verified that this actually does work.

Also just noticed a few other coverage related things:

1. Tests are missing for the text box input for the sliders in the event marker/rotation target trees. Definitely something that should be covered and pretty easy to test so that will help.
2. The path painter method that paints the grid isn't covered, just setting the show grid setting to true in split_path_editor_test/split_choreo_path_editor_test should get that.
3. The logic for adding folders/project items used to be covered in project_page but no longer is. The same logic from the old tests should be able to cover that here. Basically: tap the add button, check that it added a folder or project card.

Sorry for making you chase coverage so much lol. But having things properly covered has saved me from breaking things so many times so I think its very important.

WindingMotor commented:
Its alright lol

WindingMotor commented:
I've been very busy this past week so sorry for the delay. Most of the changes are almost implemented and i will take a look at the PR mentioned.

mjansen4857 commented:
No rush. Just as a heads up I'm currently working on a PR to update the Writerside documentation for all the changes I made but haven't documented yet. After that's merged you should update the docs for your changes in this PR. All you should have to do is update a few screenshots and do quick descriptions of additions like the new editor settings. Writerside is a free Jetbrains app or IntelliJ plugin, but the changes you'd need to make are so minor you can probably just edit the raw Markdown/replace existing images.

The documentation updates can also be done in a future PR if you'd prefer to get this merged a bit quicker. I don't really care either way as long as the docs get updated before beta which is still a ways out.

WindingMotor commented:
I would prefer to use a separate PR for the documentation, thanks!

mjansen4857 commented:
Just checking in to see how things are going with this. I have a branch ready for #690 which touches most files in the project so I want to get this merged before that if you think it'll be ready in the next few days so that I don't cause a million different merge conflicts for you. Again, no rush, just wondering if its worth holding off.

WindingMotor commented:
It should be ready tonight. I am working on the current conflicts. Thanks for your patience

WindingMotor commented:
Okay I have merged everything and ran LCOV with 84.4% of the project being covered. I looked into the issue of at #723 and could not replicate it.

mjansen4857 commented:
Ok. I'll try to keep a look out for that issue. Thanks for all the work on this.

JosephTLockwood commented:
It only appears to happen when "Recent" is the sort by method. I tested again and could not have the issue occur when sorted Alphabetically. It shouldn't be a huge deal, even if it's an issue.

WindingMotor commented:
@mjansen4857 Thanks for all your help with this PR, I learned a lot!

mjansen4857 commented:
No problem. Just a heads up I made a couple minor fixes to things I noticed after merging in #776 

Main things were some cards were darker than others and the robot details would rotate with the robot.
